File: SW.Bitween.Web/ClientApp/src/api/http/audit.ts
import type { AuditQuery, Paged, TrailEntry } from "../types";
import { get } from "./request";

interface RawAuditEntry {
  id: string;
  correlationId: string;
  sequence: number;
  occurredOn: string;
  userId: string | null;
  userDisplayName: string | null;
  entityName: string;
  entityKey: string;
  state: "Added" | "Modified" | "Deleted";
  changes: Record<string, { old: unknown; new: unknown }> | null;
}

interface SearchyResponse<T> {
  result: T[];
  totalCount: number;
}

const toEntry = (r: RawAuditEntry): TrailEntry => ({
  id: r.id,
  on: r.occurredOn,
  action: r.state,
  // The backend resolves the name on read, so a deleted account leaves the id showing
  // rather than a name that is no longer true.
  by: r.userDisplayName ?? (r.userId ? `#${r.userId}` : "System"),
  byUserId: r.userDisplayName ? (r.userId ?? undefined) : undefined,
  entityName: r.entityName,
  entityKey: r.entityKey,
  correlationId: r.correlationId,
  changes: Object.entries(r.changes ?? {}).map(([property, diff]) => ({
    property,
    old: diff.old,
    new: diff.new,
  })),
});

function buildQuery(query: AuditQuery): string {
  const params = new URLSearchParams();
  params.set("offset", String(query.offset));
  params.set("limit", String(query.limit));
  if (query.entityName) params.set("entityName", query.entityName);
  if (query.entityKey) params.set("entityKey", query.entityKey);
  if (query.userId) params.set("userId", query.userId);
  if (query.correlationId) params.set("correlationId", query.correlationId);
  if (query.from) params.set("from", query.from);
  if (query.to) params.set("to", query.to);
  return params.toString();
}

export async function searchAudit(query: AuditQuery): Promise<Paged<TrailEntry>> {
  const res = await get<SearchyResponse<RawAuditEntry>>(`/audit?${buildQuery(query)}`);
  return { result: (res.result ?? []).map(toEntry), total: res.totalCount ?? 0 };
}

export const auditMethods = { searchAudit };

File: SW.Bitween.Sdk/Model/Audit.cs
using System;
using System.Collections.Generic;

namespace SW.Bitween.Model;

/// <summary>
/// Filters for the audit trail. Every field is optional; with none set the result is the whole
/// trail, newest first.
/// </summary>
public class SearchAuditModel
{
    public int? Limit { get; set; }
    public int? Offset { get; set; }

    /// <summary>Narrows to one kind of entity, e.g. <c>Subscription</c>.</summary>
    public string EntityName { get; set; }

    /// <summary>With <see cref="EntityName"/>, the history of one row.</summary>
    public string EntityKey { get; set; }

    /// <summary>The account behind the change.</summary>
    public string UserId { get; set; }

    /// <summary>Everything one save changed, as a group.</summary>
    public string CorrelationId { get; set; }

    public DateTime? From { get; set; }
    public DateTime? To { get; set; }
}

public class AuditEntryModel
{
    public string Id { get; set; }
    public string CorrelationId { get; set; }
    public int Sequence { get; set; }
    public DateTime OccurredOn { get; set; }

    public string UserId { get; set; }

    /// <summary>
    /// The account's display name at the time it is read, or null for a change made with no signed-in
    /// user — a bus consumer or a scheduled job. Resolved on read rather than stored, so it is
    /// blank rather than wrong once an account is deleted.
    /// </summary>
    public string UserDisplayName { get; set; }

    public string EntityName { get; set; }
    public string EntityKey { get; set; }

    /// <summary><c>Added</c>, <c>Modified</c> or <c>Deleted</c>.</summary>
    public string State { get; set; }

    /// <summary>Property name to its before/after values.</summary>
    public Dictionary<string, AuditChangeModel> Changes { get; set; }
}

public class AuditChangeModel
{
    public object Old { get; set; }
    public object New { get; set; }
}
